File: src/dotnet/PolyBus/IPolyBus.cs
using PolyBus.Transport;
using PolyBus.Transport.Transactions;
using PolyBus.Transport.Transactions.Messages;
using PolyBus.Transport.Transactions.Messages.Handlers;

namespace PolyBus;

public interface IPolyBus
{
    IDictionary<string, object> Properties { get; }

    ITransport Transport { get; }

    IList<IncomingHandler> IncomingPipeline { get; }

    IList<OutgoingHandler> OutgoingPipeline { get; }

    Messages Messages { get; }

    Task<IncomingTransaction> CreateIncomingTransaction(IncomingMessage message);

    Task<OutgoingTransaction> CreateOutgoingTransaction();

    Task Send(Transaction transaction);

    Task Start();

    Task Stop();

    string Name { get; }
}

File: src/kotlin/src/main/kotlin/polybus/IPolyBus.kt
package polybus

import polybus.transport.ITransport
import polybus.transport.transactions.IncomingTransaction
import polybus.transport.transactions.OutgoingTransaction
import polybus.transport.transactions.Transaction
import polybus.transport.transactions.messages.IncomingMessage
import polybus.transport.transactions.messages.Messages
import polybus.transport.transactions.messages.handlers.IncomingHandler
import polybus.transport.transactions.messages.handlers.OutgoingHandler

interface IPolyBus {
    val properties: MutableMap<String, Any>
    var transport: ITransport
    val incomingPipeline: MutableList<IncomingHandler>
    val outgoingPipeline: MutableList<OutgoingHandler>
    val messages: Messages

    suspend fun createIncomingTransaction(message: IncomingMessage): IncomingTransaction
    suspend fun createOutgoingTransaction(): OutgoingTransaction
    suspend fun send(transaction: Transaction)
    suspend fun start()
    suspend fun stop()

    val name: String
}
